torch/csrc/jit/codegen/cuda/kernel_ir.h

```diff
@@ -1011,13 +1013,27 @@ class TORCH_CUDA_CU_API TensorIndex final : public Val {
     return indices_;
   }
 
+  const std::vector<bool>& rootDomainIndexMap() const {
+    return root_domain_index_map_;
+  }
+
   const TensorView* view() const {
     return view_;
   }
 
+  bool address() const {
+    return address_;
+  }
+
  private:
   const TensorView* view_ = nullptr;
   std::vector<Val*> indices_;
+
+  // Identify which root domain's have a non-zero index
+  std::vector<bool> root_domain_index_map_;
+
+  // Generate the address for the TensorIndex
+  bool address_;
```

Comment:
This seems like a hack. Why not adding `UnaryOpType::AddressOf`?

Reply:
I modified `TensorIndex` based on Christian's suggestion. He avoided adding a new `UnaryOpType` before because he didn't want them exposed to the user. e.g., `UnaryOpType::VectorizedSet` You can ask him which he prefers in this case.

Reply:
@csarofeen

That doesn't seem to be a concern since it's only for KIR. It's not possible to hide it from the user since the enum is shared between the two IRs, but we can error out when used in user programs.

It's fairly natural that some ops are only supported in either of the IRs, so if we really want to hide unsupported ops, we should define a separate enum for each IR.

`&` is a standard unary op in CUDA, and I don't see any reason we should not have that in KIR.

Reply:
The unary ops I really didn't like were the vectorized read and vectorized write one. I agree this unary op is a bit of a concern, but it does seem to make sense to have in the kernel ir. Since we don't have a separate enum for the kernel ir, what we should do is make sure it's not accessible in arith.cpp/.h. I think what I recommended to you Ryan was an address type, but it makes sense that it wouldn't be helpful without a matching unary op that can take in a tensor index and return an address (can just be an int64_t). As long as it's properly protected in arith.cpp to prevent users from ever using this unary op enum, then I think we can go ahead with it.

Reply:
@rdspring1 Thanks for adding the unary op. Are you also adding an address type?

Reply:
Not currently. You mean like `ptrdiff_t` => `class Pointer` in `kir` namespace.

`int64_t` is the data type for size, stride, and indices, so I used that for the pointer address.

Reply:
Yes, that's what I was thinking of.

At this point, we don't directly dereference those taken addresses but just calculate offsets, so just `int64_t` would be probably just fine. We may want to reconsider in the future if more complex pointer arithmetic is involved.